### ISSUE TITLE
PR #389 greatly slows 8-bit LoRA training (via bitsandbytes)

### ISSUE THREAD

**Issue by rationalism** [huggingface/peft#422]
This PR: 

https://github.com/huggingface/peft/pull/389

greatly slows 8-bit fine-tuning of LLaMA-33b on my machine, presumably because ops are being done in 32-bit instead of 16-bit. (I didn't experience any apparent instability or other problems on the old version.) I confirmed this by doing benchmarking on that commit (`12aebc56f9`) vs. the previous commit (`8e53e16005`), average iteration times go from 8.85 seconds to 13.17 seconds (49% longer). `bitsandbytes` also now prints out this error message:

`/home/alyssa/anaconda3/envs/lm_fun/lib/python3.10/site-packages/bitsandbytes/autograd/_functions.py:298: UserWarning: MatMul8bitLt: inputs will be cast from torch.float32 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")`
  
which is presumably related. Reverting to previous commit fixes the issue. I am running PyTorch 2.0.0, HuggingFace Transformers 4.28.1, Ubuntu 22.04 and CUDA 11.7.

@pacman100

**Comment by Splo2t** [comment]
I compared the number of float16 weights that were converted to float32 between PEFT versions 0.2.0 and 0.3.0 (including PR #389 ). In version 0.2.0, no weights were changed, but in version 0.3.0, all weights that were not 8-bit quantized were typecaste.

'''python
transformer.h.0.ln_1.weight, torch.float16
transformer.h.0.ln_1.bias, torch.float16
transformer.h.0.mlp.fc_in.bias, torch.float16
transformer.h.0.mlp.fc_out.bias, torch.float16

lm_head.weight, torch.float16
lm_head.bias, torch.float16
'''

When the weights above were trained as fp16 without typecasting to fp32 in version 0.3.0, the same training time was recorded as in the previous 0.2.0 version.

Tensor operations are typically faster with fp16 than with fp32, but they have a trade-off in accuracy. Indeed, several issues (#350 , #236 ...) reported cases where the loss did not converge. As a result, it is believed that PR #389 opted for typecasting to float32, even at the cost of reduced speed.

**Comment by rationalism** [comment]
Would it be possible to have method parameters to specify (`fp32`, `fp16`, `bf16`)? `bf16` has been useful for stability on supported GPUs, IME

**Comment by Splo2t** [comment]
I have tried modifying the internal library and tested this feature. However, it seems that there is currently no option for that. I have submitted a PR in relation to this matter.

**Comment by ManuelFay** [comment]
I did notice significant slowdowns and this error message on the Llama model (22s to 50s per batch). I did not dive deep into why but if this PR is the root cause, might be interesting to dive into it  a bit more !

**Comment by pacman100** [comment]
Hello, please see this: https://github.com/huggingface/peft/pull/423#issuecomment-1582209118

Closing this issue inline with it.
